store/tikv/range_task.go

```diff
@@ -0,0 +1,298 @@
+	for i := 0; i < s.concurrency; i++ {
+		w := s.createWorker(taskCh, &wg)
+		wg.Add(1)
+		go w.run(ctx, cancel)
+	}
+
+	defer func() {
+		close(taskCh)
+		statLogTicker.Stop()
+		cancel()
+		wg.Wait()
+	}()
+
+	startTime := time.Now()
+
+	// Iterate all regions and send each region's range as a task to the workers.
+	key := startKey
+	for {
+		select {
+		case <-ctx.Done():
+			return errors.Trace(ctx.Err())
+		case <-statLogTicker.C:
+			logutil.Logger(ctx).Info("range task in progress",
+				zap.String("name", s.name),
+				zap.Binary("startKey", startKey),
+				zap.Binary("endKey", endKey),
+				zap.Duration("costTime", time.Since(startTime)),
+				zap.Int32("completedRegions", s.CompletedRegions()))
+		default:
+		}
+
+		bo := NewBackoffer(ctx, locateRegionMaxBackoff)
+
+		loc, err := s.store.GetRegionCache().LocateKey(bo, key)
+		if err != nil {
+			logutil.Logger(ctx).Info("range task failed",
+				zap.String("name", s.name),
+				zap.Binary("startKey", startKey),
+				zap.Binary("endKey", endKey),
+				zap.Error(err))
+			return errors.Trace(err)
+		}
+		r := &kv.KeyRange{
+			StartKey: key,
+			EndKey:   loc.EndKey,
+		}
+
+		isLast := len(r.EndKey) == 0 || (len(endKey) > 0 && bytes.Compare(r.EndKey, endKey) >= 0)
+		// Let r.EndKey = min(endKey, loc.EndKey)
+		if isLast {
+			r.EndKey = endKey
+		}
+
+		taskCh <- r
+
+		if isLast {
+			break
+		}
+
+		key = r.EndKey
+	}
+
+	logutil.Logger(ctx).Info("range task finished",
+		zap.String("name", s.name),
+		zap.Binary("startKey", startKey),
+		zap.Binary("endKey", endKey))
+
+	return nil
+}
+
+// createWorker creates a worker that can process tasks from the given channel.
+func (s *RangeTaskRunner) createWorker(taskCh chan *kv.KeyRange, wg *sync.WaitGroup) *rangeTaskWorker {
+	return &rangeTaskWorker{
+		name:       s.name,
+		store:      s.store,
+		handler:    s.handler,
+		maxBackoff: s.maxBackoff,
+		taskCh:     taskCh,
+		wg:         wg,
+
+		totalCompletedRegions: &s.completedRegions,
+	}
+}
+
+// CompletedRegions returns how many regions has been sent requests.
+func (s *RangeTaskRunner) CompletedRegions() int32 {
+	return atomic.LoadInt32(&s.completedRegions)
+}
+
+// rangeTaskWorker is used by RangeTaskRunner to process tasks concurrently.
+type rangeTaskWorker struct {
+	name       string
+	store      Storage
+	handler    RangeTaskHandler
+	maxBackoff int
+	taskCh     chan *kv.KeyRange
+	wg         *sync.WaitGroup
+
+	completedRegions int32
+
+	totalCompletedRegions *int32
+}
+
+// run runs the worker. It collects all objects from `w.taskCh` and process them one by one.
+func (w *rangeTaskWorker) run(ctx context.Context, cancel context.CancelFunc) {
+	// Each task might take very short time to finish, so if we update the statistics immediately after sending a
+	// request to a region, the contention of atomic operations might reduce the performance. So it's better to
+	// calculate only the current thread and update it to the RangeTaskRunner.
```

Comment:
I think it will be fine to update after each task. Because every task evolves rpc call(s), which will cost much more time than atomic operation